src/Dependencies/Collections/Internal/ThrowHelper.cs

```diff
@@ -0,0 +1,343 @@
+        }
+
+        private static KeyNotFoundException GetKeyNotFoundException(object? key)
+        {
+            return new KeyNotFoundException(string.Format(SR.Arg_KeyNotFoundWithKey, key));
+        }
+
+        private static ArgumentOutOfRangeException GetArgumentOutOfRangeException(ExceptionArgument argument, ExceptionResource resource)
+        {
+            return new ArgumentOutOfRangeException(GetArgumentName(argument), GetResourceString(resource));
+        }
+
+        // Allow nulls for reference types and Nullable<U>, but not for value types.
+        // Aggressively inline so the jit evaluates the if in place and either drops the call altogether
+        // Or just leaves null test and call to the Non-returning ThrowHelper.ThrowArgumentNullException
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void IfNullAndNullsAreIllegalThenThrow<T>(object? value, ExceptionArgument argName)
+        {
+            // Note that default(T) is not equal to null for value types except when T is Nullable<U>.
+            if (!(default(T) == null) && value == null)
+                ThrowHelper.ThrowArgumentNullException(argName);
+        }
+
+        private static string GetArgumentName(ExceptionArgument argument)
+        {
+            switch (argument)
+            {
+                case ExceptionArgument.dictionary:
+                    return "dictionary";
+                case ExceptionArgument.array:
+                    return "array";
+                case ExceptionArgument.key:
+                    return "key";
+                case ExceptionArgument.value:
+                    return "value";
+                case ExceptionArgument.startIndex:
+                    return "startIndex";
+                case ExceptionArgument.index:
+                    return "index";
+                case ExceptionArgument.capacity:
+                    return "capacity";
+                case ExceptionArgument.collection:
+                    return "collection";
+                case ExceptionArgument.item:
+                    return "item";
+                case ExceptionArgument.converter:
+                    return "converter";
+                case ExceptionArgument.match:
+                    return "match";
+                case ExceptionArgument.count:
+                    return "count";
+                case ExceptionArgument.action:
+                    return "action";
+                case ExceptionArgument.comparison:
+                    return "comparison";
+                case ExceptionArgument.length:
+                    return "length";
+                case ExceptionArgument.destinationArray:
+                    return "destinationArray";
+                default:
+                    Debug.Fail("The enum value is not defined, please check the ExceptionArgument Enum.");
+                    return "";
```

Comment:
Consider making these just throw for the default case and simplifying to a switch expression.

Reply:
This is verbatim from [dotnet/runtime](https://github.com/dotnet/runtime/blob/f56ef95669c115510c0b6dc294634d932fd4e49b/src/libraries/System.Private.CoreLib/src/System/ThrowHelper.cs#L713-L714)